File: scripts/fetch_ratings.py
"""
用 Playwright 爬 Google Maps 評分和評論數
"""
from playwright.sync_api import sync_playwright
import json
import time
import sys
import io

sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

shops = [
    {"city": "taipei", "name": "冷水坑", "county": "台北", "area": "陽明山"},
    {"city": "newtaipei", "name": "碧潭", "county": "新北", "area": "新店"},
    {"city": "taichung", "name": "審計新村", "county": "台中", "area": "西區"},
    {"city": "tainan", "name": "神農街", "county": "台南", "area": "中西區"},
    {"city": "kaohsiung", "name": "橋頭糖廠", "county": "高雄", "area": "橋頭"},
]

results = []

with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)
    page = browser.new_page()

    for shop in shops:
        try:
            query = f"{shop['name']} {shop['county']}{shop['area']}"
            print(f"搜尋: {query}")
            url = f"https://www.google.com/maps/search/{query}"
            page.goto(url, wait_until="domcontentloaded", timeout=30000)
            time.sleep(5)

            content = page.inner_text("body")
            lines = content.split("\n")

            rating = None
            reviews = None

            for i, line in enumerate(lines):
                line = line.strip()
                # Look for rating pattern like "4.3" standalone
                if rating is None:
                    import re
                    m = re.match(r'^(\d\.\d)$', line)
                    if m:
                        rating = float(m.group(1))

                # Look for reviews count
                if reviews is None:
                    m = re.match(r'^([\d,]+)\s*(則評論|則Google評論|reviews|篇評論)', line)
                    if m:
                        reviews = int(m.group(1).replace(",", ""))
                    else:
                        m = re.match(r'^\(([\d,]+)\)\s*(則評論|reviews)', line)
                        if m:
                            reviews = int(m.group(1).replace(",", ""))

            result = {**shop, "rating": rating, "reviews": reviews}
            results.append(result)
            print(f"  -> rating={rating}, reviews={reviews}")

        except Exception as e:
            result = {**shop, "rating": None, "reviews": None, "error": str(e)}
            results.append(result)
            print(f"  -> ERROR: {e}")

        time.sleep(3)

    browser.close()

# Save results
with open("scripts/rating_results.json", "w", encoding="utf-8") as f:
    json.dump(results, f, ensure_ascii=False, indent=2)

print("\n=== Results ===")
print(json.dumps(results, ensure_ascii=False, indent=2))

File: scripts/fetch_ratings.js
const { chromium } = require('C:/Users/USER/AppData/Roaming/npm/node_modules/n8n/node_modules/playwright');
const fs = require('fs');

(async () => {
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();

  const shops = [
    { city: 'taipei', name: '冷水坑', county: '台北', area: '陽明山' },
    { city: 'newtaipei', name: '碧潭', county: '新北', area: '新店' },
    { city: 'taichung', name: '審計新村', county: '台中', area: '西區' },
    { city: 'tainan', name: '神農街', county: '台南', area: '中西區' },
    { city: 'kaohsiung', name: '橋頭糖廠', county: '高雄', area: '橋頭' }
  ];

  const results = [];

  for (const shop of shops) {
    try {
      const query = encodeURIComponent(shop.name + ' ' + shop.county + shop.area);
      await page.goto('https://www.google.com/maps/search/' + query, { waitUntil: 'networkidle', timeout: 20000 });
      await page.waitForTimeout(4000);

      // Try to find rating - look for spans with X.X pattern
      const ratingText = await page.evaluate(() => {
        const allSpans = document.querySelectorAll('span');
        for (const span of allSpans) {
          const text = span.textContent.trim();
          if (/^\d\.\d$/.test(text)) return text;
        }
        // Try aria-label with star info
        const els = document.querySelectorAll('[aria-label]');
        for (const el of els) {
          const label = el.getAttribute('aria-label') || '';
          if (/[\d.]+\s*(星|star)/i.test(label)) {
            const match = label.match(/([\d.]+)/);
            if (match) return match[1];
          }
        }
        return null;
      });

      // Try to find reviews count
      const reviewsText = await page.evaluate(() => {
        const allSpans = document.querySelectorAll('span');
        for (const span of allSpans) {
          const text = span.textContent.trim();
          if (/[\d,]+\s*(則評論|reviews|篇評論|則Google評論)/.test(text)) return text;
        }
        // Also try divs
        const allDivs = document.querySelectorAll('div');
        for (const div of allDivs) {
          const text = div.textContent.trim();
          if (/^[\d,]+\s*(則評論|reviews)$/.test(text)) return text;
        }
        return null;
      });

      let rating = null;
      let reviews = null;

      if (ratingText) {
        const match = ratingText.match(/([\d.]+)/);
        if (match) rating = parseFloat(match[1]);
      }

      if (reviewsText) {
        const match = reviewsText.match(/([\d,]+)/);
        if (match) reviews = parseInt(match[1].replace(/,/g, ''));
      }

      results.push({ ...shop, rating, reviews });
      console.log(shop.name + ': rating=' + rating + ', reviews=' + reviews);
    } catch(e) {
      results.push({ ...shop, rating: null, reviews: null, error: e.message });
      console.log(shop.name + ': ERROR - ' + e.message);
    }
    await page.waitForTimeout(3000);
  }

  // Write results to a temp file
  fs.writeFileSync(__dirname + '/rating_results.json', JSON.stringify(results, null, 2), 'utf8');
  console.log('\nResults saved to scripts/rating_results.json');
  console.log(JSON.stringify(results, null, 2));

  await browser.close();
})();